File: src/00_sample_lists.py
"""Build the training sample lists for ibeji.

Matches GEUVADIS expression samples to 1000 Genomes phase 3 genotypes, then
writes PLINK 2 --keep files for the European and Yoruba sets plus five random
European downsamples matched to the Yoruba sample size.
"""
from pathlib import Path

import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
RAW = ROOT / "data" / "raw"
OUT = ROOT / "data" / "processed" / "samples"
OUT.mkdir(parents=True, exist_ok=True)

EUR_POPS = {"CEU", "FIN", "GBR", "TSI"}
N_DOWNSAMPLE_REPS = 5

expr_header = pd.read_csv(
    RAW / "GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz", sep="\t", nrows=0
)
expr_samples = list(expr_header.columns[4:])

psam = pd.read_csv(RAW / "phase3_corrected.psam", sep="\t").set_index("#IID")
matched = [s for s in expr_samples if s in psam.index]
missing = [s for s in expr_samples if s not in psam.index]

pops = psam.loc[matched, "Population"]
eur = sorted(pops[pops.isin(EUR_POPS)].index)
yri = sorted(pops[pops == "YRI"].index)


def write_keep(name: str, ids: list[str]) -> None:
    pd.DataFrame({"#IID": ids}).to_csv(OUT / f"{name}.txt", sep="\t", index=False)


write_keep("ALL", sorted(eur + yri))
write_keep("EUR358", eur)
write_keep("YRI87", yri)

rows = []
for rep in range(1, N_DOWNSAMPLE_REPS + 1):
    draw = sorted(pd.Series(eur).sample(n=len(yri), random_state=rep))
    write_keep(f"EUR87_r{rep}", draw)
    comp = pops.loc[draw].value_counts().to_dict()
    rows.append({"set": f"EUR87_r{rep}", "n": len(draw), **comp})

summary = pd.DataFrame(
    [{"set": "EUR358", "n": len(eur), **pops.loc[eur].value_counts().to_dict()},
     {"set": "YRI87", "n": len(yri), "YRI": len(yri)}] + rows
).fillna(0)
summary.to_csv(OUT / "set_composition.tsv", sep="\t", index=False)

print(f"expression samples: {len(expr_samples)}; matched to 1KG phase 3: {len(matched)}")
print(f"missing from phase 3 ({len(missing)}): {', '.join(missing)}")
print(summary.to_string(index=False))

File: src/00_sample_lists.R
## Build the training sample lists for ibeji.
## Matches GEUVADIS expression samples to 1000 Genomes phase 3 genotypes, then writes
## PLINK 2 --keep files for the European and Yoruba sets plus five European draws.
##
## This will not reproduce the existing downsamples. R's sample() and pandas'
## .sample(random_state=rep) are different generators, so it draws five valid but
## different European sets and overwrites the lists the trained models depend on.
## ALL, EUR358 and YRI87 are deterministic and do match.

suppressPackageStartupMessages(library(data.table))

find_root <- function() {
  a <- commandArgs(trailingOnly = FALSE)
  f <- sub("^--file=", "", a[grep("^--file=", a)])
  d <- if (length(f)) dirname(normalizePath(f[1L])) else getwd()
  normalizePath(file.path(d, ".."), mustWork = FALSE)
}
ROOT <- find_root()
RAW  <- file.path(ROOT, "data", "raw")
OUT  <- file.path(ROOT, "data", "processed", "samples")
dir.create(OUT, recursive = TRUE, showWarnings = FALSE)

EUR_POPS          <- c("CEU", "FIN", "GBR", "TSI")
N_DOWNSAMPLE_REPS <- 5L

## Read only the header of the expression matrix: the sample identifiers begin at
## column 5, after the four annotation columns.
expr_header  <- fread(file.path(RAW, "GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz"),
                      sep = "\t", nrows = 0L)
expr_samples <- names(expr_header)[-seq_len(4L)]

psam <- fread(file.path(RAW, "phase3_corrected.psam"), sep = "\t")
setnames(psam, sub("^#", "", names(psam)))
setkey(psam, IID)

matched <- expr_samples[expr_samples %in% psam$IID]
missing <- expr_samples[!expr_samples %in% psam$IID]

pops <- psam[match(matched, IID), Population]
names(pops) <- matched
eur <- sort(names(pops)[pops %in% EUR_POPS])
yri <- sort(names(pops)[pops == "YRI"])

write_keep <- function(name, ids) {
  d <- data.table(IID = ids)
  setnames(d, "IID", "#IID")
  fwrite(d, file.path(OUT, paste0(name, ".txt")), sep = "\t")
}

write_keep("ALL",    sort(c(eur, yri)))
write_keep("EUR358", eur)
write_keep("YRI87",  yri)

rows <- list()
for (rep in seq_len(N_DOWNSAMPLE_REPS)) {
  set.seed(rep)
  draw <- sort(sample(eur, length(yri)))
  write_keep(sprintf("EUR87_r%d", rep), draw)
  comp <- as.list(table(pops[draw]))
  rows[[rep]] <- c(list(set = sprintf("EUR87_r%d", rep), n = length(draw)), comp)
}

summary_rows <- c(
  list(c(list(set = "EUR358", n = length(eur)), as.list(table(pops[eur])))),
  list(list(set = "YRI87", n = length(yri), YRI = length(yri))),
  rows)
summary <- rbindlist(summary_rows, fill = TRUE)
for (j in names(summary)) if (is.numeric(summary[[j]])) set(summary, which(is.na(summary[[j]])), j, 0)
fwrite(summary, file.path(OUT, "set_composition.tsv"), sep = "\t")

cat(sprintf("expression samples: %d; matched to 1KG phase 3: %d\n", length(expr_samples), length(matched)))
cat(sprintf("missing from phase 3 (%d): %s\n", length(missing), paste(missing, collapse = ", ")))
print(summary)
